src/hevm/src/EVM/Transaction.hs

```diff
@@ -8,53 +8,75 @@ import EVM.FeeSchedule
 import EVM.Precompiled (execute)
 import EVM.RLP
 import EVM.Symbolic (forceLit)
-import EVM.Types (keccak)
 import EVM.Types
 
 import Control.Lens
 
 import Data.Aeson (FromJSON (..))
 import Data.ByteString (ByteString)
-import Data.Map (Map)
+import Data.Map (Map, keys)
+import Data.Set (fromList)
 import Data.Maybe (fromMaybe, isNothing, isJust)
 
 import qualified Data.Aeson        as JSON
 import qualified Data.Aeson.Types  as JSON
 import qualified Data.ByteString   as BS
 import qualified Data.Map          as Map
 
-data Transaction = Transaction
-  { txData     :: ByteString,
+data AccessListEntry = AccessListEntry {
+  accessAddress :: Addr,
+  accessStorageKeys :: [W256]
+} deriving Show
+
+data Transaction = Transaction {
+    txData     :: ByteString,
     txGasLimit :: W256,
     txGasPrice :: W256,
     txNonce    :: W256,
     txR        :: W256,
     txS        :: W256,
     txToAddr   :: Maybe Addr,
     txV        :: W256,
-    txValue    :: W256
-  } deriving Show
+    txValue    :: W256,
+    txType     :: Maybe W256,
+    txAccessList :: Maybe [AccessListEntry]
+} deriving Show
+
+-- utility function for getting a more useful representation of accesslistentries
+-- duplicates only matter for gas computation
+-- ugly! could use a review....
+txAccessMap :: Transaction -> Map Addr [W256]
+txAccessMap tx = maybe mempty ((Map.fromListWith (++)) . makeTups) $ txAccessList tx
+  where makeTups = map (\ale -> (accessAddress ale, accessStorageKeys ale))
 
 ecrec :: W256 -> W256 -> W256 -> W256 -> Maybe Addr
-ecrec e v r s = (num . word) <$> EVM.Precompiled.execute 1 input 32
+ecrec v r s e = num . word <$> EVM.Precompiled.execute 1 input 32
   where input = BS.concat (word256Bytes <$> [e, v, r, s])
 
 sender :: Int -> Transaction -> Maybe Addr
-sender chainId tx = ecrec hash v' (txR tx) (txS tx)
-  where hash = keccak $ signingData chainId tx
+sender chainId tx = hash >>= (ecrec v' (txR tx) (txS tx))
+  where hash = keccak <$> signingData chainId tx
         v    = txV tx
         v'   = if v == 27 || v == 28 then v
                else 28 - mod v 2
 
-signingData :: Int -> Transaction -> ByteString
+signingData :: Int -> Transaction -> Maybe ByteString
 signingData chainId tx =
-  if v == (chainId * 2 + 35) || v == (chainId * 2 + 36)
-  then eip155Data
-  else normalData
+  case txType tx of
+    Nothing -> (if v == (chainId * 2 + 35) || v == (chainId * 2 + 36)
+      then Just eip155Data
+      else Just normalData)
+    Just 0x01 -> Just eip2930Data
+    _ -> Nothing
   where v          = fromIntegral (txV tx)
         to'        = case txToAddr tx of
           Just a  -> BS $ word160Bytes a
           Nothing -> BS mempty
+        accessList = (maybe [] id) . txAccessList $ tx
+        rlpAccessList = BS.concat $ map (\accessEntry ->
+          rlpList [BS $ word160Bytes (accessAddress accessEntry), 
+          BS $ rlpList $ map rlpWord256 $ accessStorageKeys accessEntry
+          ]) accessList
```

Comment:
```suggestion
        rlpAccessList = rlpList $ map (\accessEntry ->
          BS $ rlpList [BS $ word160Bytes (accessAddress accessEntry)
                     ,BS $ rlpList $ map rlpWord256 $ accessStorageKeys accessEntry]
          ) accessList
```

Reply:
needs confirmation, but I think something like this would make more sense to get the rlp encoding right

Reply:
I think this is still wrong, I'm still getting a sender hash that doesn't exist in the prestate of the test. I'll mess with it and see what happens.